src/java.base/share/classes/java/io/DataInputStream.java

```diff
@@ -573,33 +579,38 @@ public final String readUTF() throws IOException {
      */
     public static final String readUTF(DataInput in) throws IOException {
         int utflen = in.readUnsignedShort();
-        byte[] bytearr = null;
-        char[] chararr = null;
-        if (in instanceof DataInputStream dis) {
+        byte[] bytearr;
+        DataInputStream dis = null;
+        if (in instanceof DataInputStream) {
+            dis = (DataInputStream)in;
             if (dis.bytearr.length < utflen) {
-                dis.bytearr = new byte[utflen*2];
-                dis.chararr = new char[utflen*2];
+                dis.bytearr = new byte[utflen << 1];
             }
-            chararr = dis.chararr;
             bytearr = dis.bytearr;
         } else {
             bytearr = new byte[utflen];
-            chararr = new char[utflen];
         }
-
-        int c, char2, char3;
-        int count = 0;
-        int chararr_count=0;
-
         in.readFully(bytearr, 0, utflen);
 
-        while (count < utflen) {
-            c = (int) bytearr[count] & 0xff;
-            if (c > 127) break;
-            count++;
-            chararr[chararr_count++]=(char)c;
+        int count;
+        char[] chararr;
+        if (utflen > 7) {
+            // When there's a chunk of ASCII 0-127, the modified UTF-8 stream can use a similar
+            // fast-path as regular UTF-8.
+            count = JLA.countPositives(bytearr, 0, utflen);
+            if (count == utflen) {
+                // For ASCII ISO-8859-1 is equivalent to UTF-8, while avoiding a redundant
+                // scan
+                return new String(bytearr, 0, utflen, StandardCharsets.ISO_8859_1);
```

Comment:
Not sure this is correct.
If `bytearr` contains some `(byte)0`, that is, if `in` is malformed, this doesn't throw `UTFDataFormatException`, but it should: modified UTF-8 cannot contain zeros.

Reply:
While properly encoded modified UTF-8 strings won't have embedded zeros (`\u0000` will be encoded as `0xC0, 0x80`) the decoding routines in `DataInputStream` and `ObjectInputStream` allows them and does not throw an exception if an embedded zero is encountered. This PR does not change semantics here AFAICT. If you think we need to be stricter in these decoders that could be done as a separate RFE and I'll put this on hold.

Reply:
Ah OK.

I didn't check the current code, only the proposed one.
Although the specification clearly states that the method should throw, if the current code does not throw on zeros, then it makes sense that the proposed one shouldn't either.

Reply:
In fact, the implementations of `readUTF*()` in `DataInputStream` and `ObjectInputStream` are much more lenient than that. They also accept ASCII characters that are encoded with 2 bytes instead of 1. There's no check that the encoding is "minimal length".

Reply:
This is according to `DataInput` specification.
So what `UTFDataFormatException` means is kind of ambiguous.